### PR TITLE
SFrame parameter negotiation

### PR DESCRIPTION
**Pull request by bifurcation**
Implements SFrame parameter negotiation as specified in the [corresponding spec PR](https://github.com/bifurcation/sframe-mls/pull/3).  

Note that this is the first extension we have that is copied into the group state.  Currently application-defined extensions cannot be auto-copied over from the GroupInfo to the GroupContext.  The application would have to manually call `state.extensions().add(custom_extension)`, and even that is currently impossible because the `extensions()` method is `const`.  We might want to consider having a better story for application-defined extensions.

### CODE REVIEW

File: test/state.cpp [hunk 1]
```diff
@@ -94,11 +94,65 @@ TEST_CASE_FIXTURE(StateTest, "Two Person")
...
+  // Create the initial state of the group
+  auto first0 = State{ group_id,          suite,           init_privs[0],
+                       identity_privs[0], key_packages[0], group_extensions };
+
+  // Get a KeyPackage from the second member, and verify compatiblity
+  auto compatible_capabilities = SFrameCapabilities{ { 1, 2, 3, 4 } };
+  REQUIRE(compatible_capabilities.compatible(specified_params));
+
+  auto incompatible_capabilities = SFrameCapabilities{ { 2, 3, 4 } };
+  REQUIRE_FALSE(incompatible_capabilities.compatible(specified_params));
```
Comment (suhasHere) [inline test/state.cpp:113]: isn't 4 common between the lists here..

File: test/state.cpp [hunk 2]
```diff
@@ -94,11 +94,65 @@ TEST_CASE_FIXTURE(StateTest, "Two Person")
...
+
+  auto key_package_extensions = ExtensionList{};
+  key_package_extensions.add(compatible_capabilities);
+  key_packages[1].sign(identity_privs[1],
+                       KeyPackageOpts{ key_package_extensions });
+
+  auto decoded_capabilities =
+    key_packages[1].extensions.find<SFrameCapabilities>();
+  REQUIRE(!!decoded_capabilities);
```
Comment (suhasHere) [inline test/state.cpp:122]: REQUIRE_TRUE

File: test/state.cpp [hunk 3]
```diff
@@ -94,11 +94,65 @@ TEST_CASE_FIXTURE(StateTest, "Two Person")
...
+
+  auto group = std::vector<State>{ first1, second0 };
+  verify_group_functionality(group);
+
+  // Check that both participants have the  correct SFrame parameters
+  auto first_params = first1.extensions().find<SFrameParameters>();
+  auto second_params = second0.extensions().find<SFrameParameters>();
+  REQUIRE(!!first_params);
+  REQUIRE(!!second_params);
+  REQUIRE(opt::get(first_params) == specified_params);
```
Comment (suhasHere) [inline test/state.cpp:143]: REQUIRE_TRUE

**Review by suhasHere — COMMENTED** [review]
one clarification question in the test. Otherwise looks good to me

File: test/state.cpp (hunk 1 above)
Comment (bifurcation) [reply test/state.cpp:113]: No, `4` is the value for `epoch_bits`.  I'll change it to be clearer, though.

**Review by suhasHere — APPROVED** [review]

**Pull request merged**